source/fastq/fastq.d

```diff
@@ -0,0 +1,341 @@
+/**
+ * High performance FASTQ library using iopipe
+ *
+ * Format:
+ * Excerpt from https://en.wikipedia.org/wiki/FASTQ_format
+
+ * FASTQ file normally uses four lines per sequence.
+ * Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line).
+ * Line 2 is the raw sequence letters.
+ * Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again.
+ * Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence.
+
+ * A FASTQ file containing a single sequence might look like this:
+ * @SEQ_ID
+ * GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+ * +
+ * !''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF>>>>>>CCCCCCC65
+
+ * The character '!' represents the lowest quality while '~' is the highest.
+ * Here are the quality value characters in left-to-right increasing order of
+ * quality (ASCII):
+ * !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~
+
+ *
+ * Some characteristics:
+ * - SEQ_ID is @[[:alphanumeric:]]. Where '@' marks the entry start. More description about Illumina header line below
+ * - Entry ID and header is a single line, which does not contain newline characters.
+ * - Sequence under the header line is [ATCGN]* (Perl regex).
+ * - A fastq file can be plain-text or gzip compressed.
+
+ * Illumina Header Line:
+ * @EAS139:136:FC706VJ:2:2104:15343:197393 1:Y:18:ATCACG
+
+ * EAS139	the unique instrument name
+ * 136	the run id
+ * FC706VJ	the flowcell id
+ * 2	flowcell lane
+ * 2104 	tile number within the flowcell lane
+ * 15343 	'x'-coordinate of the cluster within the tile
+ * 197393 	'y'-coordinate of the cluster within the tile
+ * 1 	the member of a pair, 1 or 2 (paired-end or mate-pair reads only)
+ * Y 	Y if the read is filtered, N otherwise
+ * 18 	0 when none of the control bits are on, otherwise it is an even number
+ * ATCACG 	index sequence. It is sample number if it is a number
+ *
+ */
+
+module fastaq.fastq;
+
+import iopipe.traits;
+import iopipe.textpipe;
+private import std.traits;
+private import std.range.primitives;
+private import std.algorithm : find, splitter, filter;
+private import std.conv: to;
+private import std.ascii: isWhite;
+private import std.string : stripLeft, stripRight, strip;
+private import std.experimental.logger;
+
+struct BufRef
+{
+  // position within the buffer of the starting reference
+  size_t pos;
+  // length of the reference
+  size_t length;
+  auto value(B)(B buf)
+  {
+    assert(pos <= buf.length);
+    assert(pos + length <= buf.length);
+    return buf[pos .. pos + length];
+  }
```

Comment:
This is identical to the one in fasta. Move this to a common file.

Reply:
Done!